### PR TITLE
CASSANDRA-19344 Range movements involving transient replicas must safely enact changes to read and write replica sets

### PR DESCRIPTION
**Pull request by beobal**
When an instance moves from a transient to a full replica for a given range, it must begin acting as a full replica for writes before it does so for reads. Otherwise, consistency can be violated as data streamed to the instance early in the operation can be removed by cleanup if it occurs before the instance assumes responsibility for full writes. Also, coordinators will route read requests to instances which may not have received all preceding writes, causing unnecessary read repair or potentially inconsistent results. 

The root cause of the flaky test failure which originally produced this issue was that in `TransientRangeMovementTest::testRemoveNode` cleanup happened to run on one instance before it had enacted the final step of the removal operation, leading it to remove more data than it should.

### CODE REVIEW

File: src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java [hunk 1]
```diff
@@ -307,17 +307,51 @@ public Builder withReplicaGroup(VersionedEndpoints.ForRange replicas)
...
 
+        /**
+         * Combine two replica groups, assuming one is the current group and the other is the proposed.
+         * During range movements this is used when calculating the maximal placement, which combines the current and
+         * future replica groups. This special cases the merging of two replica groups to make sure that when a replica
+         * moves from transient to full, it starts to act as a FULL write replica as early as possible.
+         *
+         * Where an endpoint is present in both groups, prefer the proposed iff it is a FULL replica. During a
+         * multi-step operation (join/leave/move), we want any change from transient to full to happen as early
+         * as possible so that a replica being whose ownership is modified in this way becomes FULL for writes before it
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java]: nit: Could you rephrase this part a bit?

> a replica being whose ownership is modified in this way becomes FULL for writes before it

I am a bit thrown off by replica being part.

File: src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java [hunk 2]
```diff
@@ -307,17 +307,51 @@ public Builder withReplicaGroup(VersionedEndpoints.ForRange replicas)
...
+                Replica r2 = e2.get(e);
+                if (null == r2)          // not present in next
+                    combined.add(r1);
+                else if (r2.isFull())    // prefer replica from next, if it is moving from transient to full
+                    combined.add(r2);
+                else
+                    combined.add(r1);    // replica is moving from full to transient, or staying the same
+            });
+            // any new replicas not in prev
+            e2.forEach((e, r2) -> {
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java:362]: Would it make sense to construct a set to iterate over in the above loop?

File: src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java [hunk 3]
```diff
@@ -307,17 +307,51 @@ public Builder withReplicaGroup(VersionedEndpoints.ForRange replicas)
...
+        private EndpointsForRange combine(EndpointsForRange prev, EndpointsForRange next)
+        {
+            Map<InetAddressAndPort, Replica> e1 = prev.byEndpoint();
+            Map<InetAddressAndPort, Replica> e2 = next.byEndpoint();
+            EndpointsForRange.Builder combined = prev.newBuilder(prev.size() + next.size());
+            e1.forEach((e, r1) -> {
+                Replica r2 = e2.get(e);
+                if (null == r2)          // not present in next
+                    combined.add(r1);
+                else if (r2.isFull())    // prefer replica from next, if it is moving from transient to full
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java:356]: nit: could probably short-circuit this with 

```
if (r2 != null && r2.isFull)
  combined.add(r2);
else
  combined.add(r1);
```

If you deliberately have split these, I do nit mind leaving as-is.

File: src/java/org/apache/cassandra/tcm/ownership/PlacementTransitionPlan.java [hunk 4]
```diff
@@ -125,4 +144,78 @@ public String toString()
...
+                    // BUT we split and merge ranges so we need to check containment both ways
+                    for (Replica newReadReplica : addedReadReplicas)
+                    {
+                        if (existingWriteReplicas.contains(newReadReplica))
+                            continue;
+                        boolean contained = false;
+                        Set<Range<Token>> intersectingRanges = new HashSet<>();
+                        for (Replica writeReplica : existingWriteReplicas)
+                        {
+                            if (writeReplica.isFull() == newReadReplica.isFull() || writeReplica.isFull() && newReadReplica.isTransient())
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/ownership/PlacementTransitionPlan.java]: nit: would you mind adding brackets here to make operator order more explicit here?

File: test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java [hunk 5]
```diff
@@ -199,41 +228,28 @@ public static List<Node> peerStateToNodes(Object[][] resultset)
 
     public static class NtsReplicationFactor extends ReplicationFactor
     {
-        private final Lookup lookup = new DefaultLookup();
```
Comment (ifesdjeen) [inline test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java:202]: oh thank you for removing this! This was an artifact of integrating with a simulator, but it never was actually used in this context.

**Review by ifesdjeen — APPROVED** [review]

File: src/java/org/apache/cassandra/tcm/sequences/RemoveNodeStreams.java [hunk 6]
```diff
@@ -131,33 +131,12 @@ private static MovementMap movementMap(InetAddressAndPort leaving, ClusterMetada
...
                                            candidateBuilder.add(replica, ReplicaCollection.Builder.Conflict.NONE);
                                    });
-                                   movements.putAll(newReplica, candidateBuilder.build(), ReplicaCollection.Builder.Conflict.NONE);
+                                   EndpointsForRange sources = candidateBuilder.build();
+                                   // log if newReplica is an existing transient replica moving to a full replica
+                                   if (startWriteRemovals.get(newReplica.endpoint()).contains(newReplica.range(), false))
+                                       logger.debug("Streaming transient -> full conversion to {} from {}", newReplica.endpoint(), sources);
+                                   movements.putAll(newReplica, sources, ReplicaCollection.Builder.Conflict.NONE);
                                });
-            // and check if any replicas went from transient -> full:
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/sequences/RemoveNodeStreams.java:136]: it also happens that we do not need mid and finish delta for this method anymore. Should we remove them?

File: src/java/org/apache/cassandra/tcm/sequences/RemoveNodeStreams.java [hunk 7]
```diff
@@ -131,33 +131,12 @@ private static MovementMap movementMap(InetAddressAndPort leaving, ClusterMetada
                                        if (!replica.endpoint().equals(leaving) && !replica.endpoint().equals(newReplica.endpoint()))
                                            candidateBuilder.add(replica, ReplicaCollection.Builder.Conflict.NONE);
                                    });
-                                   movements.putAll(newReplica, candidateBuilder.build(), ReplicaCollection.Builder.Conflict.NONE);
+                                   EndpointsForRange sources = candidateBuilder.build();
+                                   // log if newReplica is an existing transient replica moving to a full replica
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/sequences/RemoveNodeStreams.java:133]: We can make a similar change in `UnbootstrapStreams` calculation of the movement map, change the mid check to 

``` 
startWriteAdditions.get(removal.endpoint()).contains(removal.range(), true);
```

File: src/java/org/apache/cassandra/tcm/transformations/PrepareJoin.java [hunk 8]
```diff
@@ -156,6 +153,12 @@ public Result execute(ClusterMetadata prev)
                                                              transitionPlan.toSplit,
                                                              startJoin, midJoin, finishJoin,
                                                              joinTokenRing, streamData);
+        if (!(this instanceof UnsafeJoin) && !prev.tokenMap.isEmpty())
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/transformations/PrepareJoin.java]: Would you be OK to move this logic to an instance method, and have implemenation in PrepareJoin, and override it in UnsafeJoin? It would be great if we could keep it polymorphic.

File: src/java/org/apache/cassandra/tcm/transformations/PrepareMove.java [hunk 9]
```diff
@@ -103,7 +104,9 @@ public Result execute(ClusterMetadata prev)
         StartMove startMove = new StartMove(nodeId, transitionPlan.addToWrites(), lockKey);
         MidMove midMove = new MidMove(nodeId, transitionPlan.moveReads(), lockKey);
         FinishMove finishMove = new FinishMove(nodeId, tokens, transitionPlan.removeFromWrites(), lockKey);
-
+        Result res = PlacementTransitionPlan.assertPreExistingWriteReplica(prev.placements, transitionPlan);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/tcm/transformations/PrepareMove.java]: Do you folks think maybe having an Option here could be useful? Or some form of polymorphic result that would visually explain the significance of `null` equivalent?

File: test/distributed/org/apache/cassandra/distributed/test/TransientRangeMovement2Test.java [hunk 10]
```diff
@@ -51,9 +59,30 @@ public void testMoveBackward() throws IOException, ExecutionException, Interrupt
...
-            cluster.get(3).nodetoolResult("move", "25").asserts().success();
+            // At the start, node1 is a TRANSIENT replica for (20,30] and FULL for (30, 40], (40,] & (,10]. When moving
+            // node3 to token 25, node1 becomes a FULL replica for (25, 40], effectively going from TRANSIENT to FULL
+            // for (25,30]. A T->F transition will always cause data for that range to be streamed to the transitioning
+            // node, which happens after StartMove and before MidMove. Running cleanup before node1 considers itself a
+            // FULL replica would remove any of the newly streamed data which is marked repaired. To avoid this, we
+            // ensure that any T->F transition is applied for writes as part of the StartMove. Have the CMS node (node1)
+            // pause before the MidMove step is committed, at which point we know that streaming has completed.
+            Callable<Epoch> pending = pauseBeforeCommit(cluster.get(1), (e) -> e instanceof PrepareMove.MidMove);
+            new Thread(() -> cluster.get(3).nodetoolResult("move", "25").asserts().success()).start();
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/test/TransientRangeMovement2Test.java]: Would you mind to add a `join` after we have finished with this thread below?

File: test/distributed/org/apache/cassandra/distributed/test/TransientRangeMovement2Test.java [hunk 11]
```diff
@@ -51,9 +59,30 @@ public void testMoveBackward() throws IOException, ExecutionException, Interrupt
...
+
+            // Before allowing node1 to proceed with committing the MidMove, instruct it to pause node1 before it
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/test/TransientRangeMovement2Test.java]: nit: This line reads a bit tautological, would maybe something like "To gate/prevent node1 from proceeding with committing MidMove, instruct it to pause before enacting it" ?

File: test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java [hunk 12]
```diff
@@ -62,18 +75,34 @@ public ReplicatedRanges replicate(List<Node> nodes)
         public abstract ReplicatedRanges replicate(Range[] ranges, List<Node> nodes);
     }
 
+    public static class DCReplicas
+    {
+        public final int totalCount;
```
Comment (ifesdjeen) [inline test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java:80]: Nit: should we try keeping API here consistent with RF?

```
    public final int allReplicas;
    public final int fullReplicas;
```

File: test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java [hunk 13]
```diff
@@ -985,20 +977,84 @@ public Diff<T> invert()
...
+        List<Replica> removals = new ArrayList<>(unfiltered.additions.size());
+        for (Replica added : unfiltered.additions)
+        {
+            // Include any new FULL replicas here. If there exists the removal of a corresponding Transient
+            // replica (i.e. a switch from T -> F), add that too. We want T -> F transitions to happen early
+            // in a multi-step operation, at the same time as new write replicas are added.
+            if (added.isFull())
+            {
+                additions.add(added);
+                for (Replica removed : unfiltered.removals)
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java]: Nit: should use same style for filtering / checking removals in both branches? 

I _think_ we can also break out from the loop here early if we did add a new entry to `removals`.

File: test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java [hunk 14]
```diff
@@ -985,20 +977,84 @@ public Diff<T> invert()
...
+        List<Replica> removals = new ArrayList<>(unfiltered.removals.size());
+        for (Replica removed : unfiltered.removals)
+        {
+            // Include any new FULL replicas here. If there exists the removal of a corresponding Transient
+            // replica (i.e. a switch from T -> F), add that too. We want T -> F transitions to happen early
+            // in a multi-step operation, at the same time as new write replicas are added.
+            if (removed.isFull())
+            {
+                removals.add(removed);
+                for (Replica added : unfiltered.additions)
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java]: Nit: I think same comment applies then to the above `additionsAndTransientToFull` code.

**Review by ifesdjeen — APPROVED** [review]

File: src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java (hunk 1 above)
Comment (krummas) [reply src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java]: think its just that extra "being" which makes this confusing, fixed

File: src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java (hunk 3 above)
Comment (krummas) [reply src/java/org/apache/cassandra/tcm/ownership/PlacementForRange.java:356]: I think I prefer the split version for the comments per-case

File: src/java/org/apache/cassandra/tcm/sequences/RemoveNodeStreams.java (hunk 7 above)
Comment (krummas) [reply src/java/org/apache/cassandra/tcm/sequences/RemoveNodeStreams.java:133]: yep, almost - this became
```
if (startWriteRemovals.get(newReplica.endpoint()).contains(newReplica.range(), false))
    logger.debug("Streaming transient -> full conversion to {} from {}", newReplica, oldReplicas.get(newReplica.range()));
movements.put(oldReplicas.get(newReplica.range()), newReplica);
```

File: src/java/org/apache/cassandra/tcm/transformations/PrepareMove.java (hunk 9 above)
Comment (krummas) [reply src/java/org/apache/cassandra/tcm/transformations/PrepareMove.java]: made `assertPreExistingWriteReplica` throw `RejectedTransformationException` instead

File: test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java (hunk 12 above)
Comment (krummas) [reply test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java:80]: think the naming is clear enough that this won't introduce any issues

File: test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java (hunk 13 above)
Comment (krummas) [reply test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java]: what did you have in mind here?

Comment (ifesdjeen) [reply test/distributed/org/apache/cassandra/distributed/test/log/PlacementSimulator.java]: I thought maybe something like this:

```
            if (added.isFull())
            {
                additions.add(added);
                Optional<Replica> removed = unfiltered.removals.stream()
                                                               .filter(r -> r.isTransient() && r.node().equals(added.node()))
                                                               .findFirst();

                if (removed.isPresent())
                    removals.add(removed.get());
            }
            else
            {
                // Conversely, when a replica transitions from F -> T, it's enacted late in a multi-step operation.
                // So only include TRANSIENT additions if there is no removal of a corresponding FULL replica.
                boolean include = unfiltered.removals.stream()
                                                     .noneMatch(removed -> removed.node().equals(added.node())
                                                                           && removed.isFull());
                if (include)
                    additions.add(added);
            }
```
